sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSQLExecution.scala

```
@@ -0,0 +1,84 @@
+import org.apache.spark.util.Utils
+
+private[sql] object SparkSQLExecution {
+
+  val EXECUTION_ID_KEY = "spark.sql.execution.id"
+
+  private val _nextExecutionId = new AtomicLong(0)
+
+  private def nextExecutionId: Long = _nextExecutionId.getAndIncrement
+
+  /**
+   * Wrap a DataFrame action to track all Spark jobs in the body so that we can connect them with
+   * an execution.
+   */
+  def withNewExecution[T](sqlContext: SQLContext, df: DataFrame)(body: => T): T = {
+    val sc = sqlContext.sparkContext
+    val oldExecutionId = sc.getLocalProperty(EXECUTION_ID_KEY)
+    try {
+      if (oldExecutionId == null) {
+        val executionId = SparkSQLExecution.nextExecutionId
+        sc.setLocalProperty(EXECUTION_ID_KEY, executionId.toString)
+        val callSite = Utils.getCallSite()
+        sqlContext.listener.onExecutionStart(
+          executionId, callSite.shortForm, callSite.longForm, df, System.currentTimeMillis())
+        val r = body
+        sqlContext.listener.onExecutionEnd(executionId, System.currentTimeMillis())
+        r
+      } else {
+        // Don't support nested `withNewExecution`. This is an example of the nested
+        // `withNewExecution`:
+        //
+        // class DataFrame {
+        //   def foo: T = withNewExecution { something.createNewDataFrame().collect() }
+        // }
+        //
+        // Note: `collect` will call withNewExecution
+        // In this case, only the "executedPlan" for "collect" will be executed. The "executedPlan"
+        // for the outer DataFrame won't be executed. So it's meaningless to create a new Execution
+        // for the outer DataFrame. Even if we track it, since its "executedPlan" doesn't run,
+        // all accumulator metrics will be 0. It will confuse people if we show them in Web UI.
+        //
+        // A real case is the `DataFrame.count` method.
+        throw new IllegalArgumentException(s"$EXECUTION_ID_KEY is already set")
```

Ah, I get it now. It's because a DF operation might call other DF operations internally (e.g. count calls collect) but the user only cares about the top level operation (i.e. count, not collect). I think the proper semantics should be to inherit the existing execution ID if it already exists. I'm not sure if I see why we should throw an exception though.

By the way, I'm confused about one thing, since you didn't wrap `count` in `withNewExecution` how did your screenshot display `count` in the call site instead of `collect`?